kernel/thor/arch/x86/vt_d.cpp

```
@@ -0,0 +1,826 @@
+
+				urgentLogger() << frg::fmt("thor: IOMMU fault {}, {} request from {:02x}:{:02x}:{:x} to 0x{:x}: {} (0x{:x})",
+					i, (flags & FaultRecording::read) ? "Read" : "Write",
+					sourceId.bus(), sourceId.device(), sourceId.function(),
+					faultRecords_[i].faultInfo.load(),
+					decodeFaultReason(reason), reason) << frg::endlog;
+			}
+		}
+
+		if(status & FaultStatus::faultOverflow) {
+			urgentLogger() << "thor: IOMMU fault overflow" << frg::endlog;
+		}
+
+		regs_.store(faultStatus, FaultStatus::stickyBits);
+
+		return IrqStatus::acked;
+	}
+
+	bool supportsPassthrough() const {
+		return ecap_ & ExtendedCapability::pt;
+	}
+
+	void enableDevice(pci::PciEntity *dev) override {
+		auto lock = frg::guard(&lock_);
+
+		auto rootEntry = &rootTable_[static_cast<uint8_t>(dev->bus)];
+		PageAccessor context;
+		frg::span<ContextTable::Entry> contextTable{};
+
+		if(!(rootEntry->entry.load() & RootTable::present)) {
+			auto contextPhys = physicalAllocator->allocate(0x1000);
+			context = {contextPhys};
+
+			memset(context.get(), 0, 0x1000);
+			flush(context.get(), 0x1000);
+			contextTable = {reinterpret_cast<ContextTable::Entry *>(context.get()), 256};
+
+			rootEntry->entry.store(RootTable::present(true) | RootTable::contextEntry(contextPhys >> 12));
+			flush(rootEntry);
+		}
+
+		if(!contextTable.size()) {
+			context = {(rootEntry->entry.load() & RootTable::contextEntry) << 12};
+			contextTable = {reinterpret_cast<ContextTable::Entry *>(context.get()), 256};
+		}
+
+		auto sourceId = SourceID{
+			static_cast<uint8_t>(dev->bus),
+			static_cast<uint8_t>(dev->slot),
+			static_cast<uint8_t>(dev->function),
+		};
+
+		auto contextEntry = &contextTable[sourceId.devfn()];
+
+		contextEntry->low.store(
+			ContextTable::present(true) |
+			ContextTable::translationType(ContextTable::TranslationType::Passthrough)
+		);
+
+		contextEntry->high.store(
+			ContextTable::addressWidth(sagaw_ - 2) |
```

For now, I would prefer to simply always use a fixed page table depth instead of making this dynamic. We can fail during initialization if our desired depth is not available.